File: backend/app/schemas/draft.py
from datetime import datetime
from typing import Any, Optional

from pydantic import Field

from app.schemas.common import ORMModel


class DraftAnswer(ORMModel):
    question_id: int = Field(gt=0)
    value: Any


class DraftUpdate(ORMModel):
    form_version: Optional[int] = Field(default=None, ge=1)
    answers: Optional[list[DraftAnswer]] = None
    current_question_id: Optional[int] = Field(default=None, gt=0)
    visited_question_ids: Optional[list[int]] = None
    started_at: Optional[datetime] = None
    completed: Optional[bool] = None

File: frontend/src/types/draft.ts
import type { AnswerValue } from "./response";

export interface DraftAnswer {
  question_id: number;
  value: AnswerValue;
}

export interface DraftResponse {
  id: number;
  form_id: number;
  slug: string;
  form_version: number;
  answers: DraftAnswer[];
  current_question_id: number;
  visited_question_ids: number[];
  started_at: string;
  last_saved_at: string;
  completed: boolean;
}

export interface UpdateDraftPayload {
  form_version: number;
  answers: DraftAnswer[];
  current_question_id: number;
  visited_question_ids: number[];
  started_at: string;
  completed?: boolean;
}
